### ISSUE TITLE
Cool! how well does it work?

### ISSUE THREAD

**Issue by dhudsmith** [LukeDitria/CNN-VAE#1]
I've been thinking about implementing a resnet style vae for images. In preparation I came across your repo -- very interesting! I'm wondering if you'd be willing to share your thoughts about how well this works? In particular, I'd be interested to understand the impact of the skip connections. Also, qualitatively, what fidelity do you see in the reconstructed images? 

Kind regards,
Hudson

**Comment by LukeDitria** [comment]
Hi Hudson,
Thanks for your message (and for reminding me that I need to update this). It works reasonably well with STL10 (using the unlabeled 100k + images for training scaled down to 64x64)
 - results below for reconstruction on a random batch from the validation set (using the same hyper-parameters as in the example code - latent vector size 512, 100 epochs ect).
Images on top are the reconstruction
![VAE_STL10_64](https://user-images.githubusercontent.com/45025318/70897112-20e68800-2046-11ea-8717-7fec294e7d47.png)
 
The perception loss in my experience makes a big difference in the level of detail of the reconstructed images, i'll update the README soon with some comparative examples. The residual/skip connections in practice allow for a deeper network and I've found that in practice when I want to increase the depth (and complexity) of a model (and have better results) I can pretty reliably (depending on the application) simply replace a single conv layer (or transpose conv) with a res up or res down block like in this model.

**Comment by dhudsmith** [comment]
Thanks for the response, Luke. These look pretty good. I remember seeing the perception loss trick in one of the Fastai courses--makes sense to me! I do wonder what the implications are on the hidden representation. The BCE + KL loss is an important part of enforcing the Gaussian priors on the hidden representation. I wonder what adding the feature loss does to that? In my particular problem, I'm very interested in studying the data in the hidden representation, so I'm wary of messing with the loss function. I will certainly steal the resnet block trick. I don't see how it can be any worse, and it may be better. Thanks again!

**Comment by LukeDitria** [comment]
Hi Hudson, 
When using the feature loss the latent representation is still only going to contain information about the input image, it's simply an addition to the reconstruction loss (BCE).
Instead of the reconstruction loss only minimizing the average (or sum) pixel error of the output image (which will result in only the low frequency components being encoded) it now also has to minimize the mean reconstruction loss of the output image after it has been filtered with some learnt "feature detectors".  These feature detectors typically "look" for edges/corners ect (aka the filtered image will have all the edges and corners "highlighted" or "exaggerated" ). Therefore in order for the VAE to minimize this loss it must preserve (and therefore encode in the representation) corners and edges ect (aka high frequency components).

**Comment by LukeDitria** [comment]
I've updated the README with the comparative results if you are interested
